Add metrics checkpoint

amitvc opened the pull request:
Added statsLogger in the DbManager class to measure successful checkpoints. Sample metrics look like this -

# TYPE herd summary
herd{success="false",quantile="0.5"} NaN
herd{success="false",quantile="0.75"} NaN
herd{success="false",quantile="0.95"} NaN
herd{success="false",quantile="0.99"} NaN
herd{success="false",quantile="0.999"} NaN
herd{success="false",quantile="0.9999"} NaN
herd{success="false",quantile="1.0"} NaN
herd_count{success="false"} 0
herd_sum{success="false"} 0.0
herd{success="true",quantile="0.5"} NaN
herd{success="true",quantile="0.75"} NaN
herd{success="true",quantile="0.95"} NaN
herd{success="true",quantile="0.99"} NaN
herd{success="true",quantile="0.999"} NaN
herd{success="true",quantile="0.9999"} NaN
herd{success="true",quantile="1.0"} NaN
herd_count{success="true"} 1
herd_sum{success="true"} 1.550816699887E12

eolivelli commented on herddb-core/src/main/java/herddb/core/TableSpaceManager.java:
```diff
@@ -1530,79 +1535,85 @@ TableSpaceCheckpoint checkpoint(boolean full, boolean pin, boolean alreadLocked)
+            LOGGER.log(Level.INFO, "{0} checkpoint finish {1} started ad {2}, finished at {3}, total time {4} ms",
+                    new Object[]{nodeId, tableSpaceName, logSequenceNumber, _logSequenceNumber, Long.toString(_stop - _start)});
+            statsLogger.getOpStatsLogger(this.tableSpaceName).registerSuccessfulEvent(_stop, TimeUnit.MILLISECONDS);
+            return new TableSpaceCheckpoint(logSequenceNumber, checkpointsTableNameSequenceNumber);
```
It is not good to have a 'return' inside a finally block, please leave it inside the main try block

eolivelli commented on herddb-core/src/main/java/herddb/core/TableSpaceManager.java:
```diff
@@ -1530,79 +1535,85 @@ TableSpaceCheckpoint checkpoint(boolean full, boolean pin, boolean alreadLocked)
+            long _stop = System.currentTimeMillis();
+            LOGGER.log(Level.INFO, "{0} checkpoint finish {1} started ad {2}, finished at {3}, total time {4} ms",
+                    new Object[]{nodeId, tableSpaceName, logSequenceNumber, _logSequenceNumber, Long.toString(_stop - _start)});
+            statsLogger.getOpStatsLogger(this.tableSpaceName).registerSuccessfulEvent(_stop, TimeUnit.MILLISECONDS);
```
It is better to create this OpStatsLogger in the constructor of TableManager, otherwise you won't see the metric until a checkpoint happens.

eolivelli replied:
We have to create a main statsLogger for the tablespace

add new instance fields:

```
// container for all of the per-tablespace metrics
final StatsLogger tablespaceStasLogger; 
final OptsStatsLogger checkpointLogger;
```

In the TableSpaceManager constructor:
```
tablespaceStasLogger= mainStatsLogger.scope(tableSpaceName); 
checkpointTimeStats  = tablespaceStasLogger.getOpStatsLogger("checkpointTime");
```

and here you will have:
`checkpointTimeStats..registerSuccessfulEvent(_stop - _start, TimeUnit.MILLISECONDS);`

eolivelli commented on herddb-core/src/main/java/herddb/core/DBManager.java:
```diff
@@ -137,6 +137,7 @@
     private final ServerConfiguration serverConfiguration;
     private ConnectionsInfoProvider connectionsInfoProvider;
     private long checkpointPeriod;
+    private final StatsLogger statsLogger;
```
can we name it 'mainStatsLogger' ?
this is the container of all of the metrics

eolivelli commented:
@diegosalvi  the error on Travis seems unrelated, it should be due to your latest change around checkpoints

```
[ERROR] Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 12.909 s <<< FAILURE! - in herddb.core.RestartTest
[ERROR] recoverAfterrPartialCheckpoint(herddb.core.RestartTest)  Time elapsed: 0.9 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.fail(Assert.java:95)
	at herddb.core.RestartTest.recoverAfterrPartialCheckpoint(RestartTest.java:681)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:379)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413)
```

amitvc commented:
@eolivelli -
Addressed code review comments. Here are the new checkpoint metrics -
# TYPE herd_checkpointTime summary
herd_checkpointTime{success="false",quantile="0.5"} NaN
herd_checkpointTime{success="false",quantile="0.75"} NaN
herd_checkpointTime{success="false",quantile="0.95"} NaN
herd_checkpointTime{success="false",quantile="0.99"} NaN
herd_checkpointTime{success="false",quantile="0.999"} NaN
herd_checkpointTime{success="false",quantile="0.9999"} NaN
herd_checkpointTime{success="false",quantile="1.0"} -Infinity
herd_checkpointTime_count{success="false"} 0
herd_checkpointTime_sum{success="false"} 0.0
herd_checkpointTime{success="true",quantile="0.5"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="0.75"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="0.95"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="0.99"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="0.999"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="0.9999"} 1.550979802604E12
herd_checkpointTime{success="true",quantile="1.0"} 1.550979802604E12
herd_checkpointTime_count{success="true"} 1
herd_checkpointTime_sum{success="true"} 1.550979802604E12
# TYPE pecora_checkpointTime summary
pecora_checkpointTime{success="false",quantile="0.5"} NaN
pecora_checkpointTime{success="false",quantile="0.75"} NaN
pecora_checkpointTime{success="false",quantile="0.95"} NaN
pecora_checkpointTime{success="false",quantile="0.99"} NaN
pecora_checkpointTime{success="false",quantile="0.999"} NaN
pecora_checkpointTime{success="false",quantile="0.9999"} NaN
pecora_checkpointTime{success="false",quantile="1.0"} NaN
pecora_checkpointTime_count{success="false"} 0
pecora_checkpointTime_sum{success="false"} 0.0
pecora_checkpointTime{success="true",quantile="0.5"} NaN
pecora_checkpointTime{success="true",quantile="0.75"} NaN
pecora_checkpointTime{success="true",quantile="0.95"} NaN
pecora_checkpointTime{success="true",quantile="0.99"} NaN
pecora_checkpointTime{success="true",quantile="0.999"} NaN
pecora_checkpointTime{success="true",quantile="0.9999"} NaN
pecora_checkpointTime{success="true",quantile="1.0"} NaN
pecora_checkpointTime_count{success="true"} 0
pecora_checkpointTime_sum{success="true"} 0.0

eolivelli commented on herddb-core/src/main/java/herddb/core/TableSpaceManager.java:
```diff
@@ -1530,79 +1539,88 @@ TableSpaceCheckpoint checkpoint(boolean full, boolean pin, boolean alreadLocked)
+            return new TableSpaceCheckpoint(logSequenceNumber, checkpointsTableNameSequenceNumber);
+
+        } finally {
+            checkpointTimeStats.registerSuccessfulEvent(_stop, TimeUnit.MILLISECONDS);
```
Last nit: it is possible that *_stop* is zero.
I suggest you to declare *_stop* inside this finally block and assign the value right before this line

amitvc replied:
Sounds good. I will update it.

amitvc replied:
@eolivelli Just pushed the change. Do we need to also calculate failed checkpoint stats? I

amitvc commented:
@eolivelli I will look at the work you do for the BookKeeper stats logger. I will learn more by reading other pull requests and playing around with the code base.
